sailor/utils/oauth_wrapper/OAuthServiceImpl.py

```diff
@@ -91,38 +92,38 @@ def fetch_endpoint_data(self, endpoint_url, method, parameters=None):
 
-    def get_access_token(self):
+    def _get_session(self, params=None):
         """
-        Return an access token.
+        Return the current active session or create a new one.
 
-        This method fetches the access_token using the client credentials used to create an instance.
-        :return: An access_token in string format
+        If a session exists, check if the session is valid and return that session.
+        Otherwise create a new session.
         """
-        service = self._get_service()
-        access_token = service.get_access_token(method='POST', decoder=json.loads, key='access_token',
-                                                data={'grant_type': 'client_credentials'})
-        return access_token
-
-    def _get_session(self, method, data):
-        """
-        Return the session object based on the client credentials.
-
-        :param method: Specifies the method type. Supported method types are 'POST', 'PUT', 'PATCH'
-        :type method: str
-        :param data: A dictionary containing the value for the
-        :type data: dict
-        """
-        service = self._get_service()
-        # the get_auth_session method of rauth does not check whether the response was 200 or not
-        # and therefore does not log a proper error message
-        session = service.get_auth_session(method=method, data=data, decoder=json.loads)
-        session.headers = {"Accept": "application/json"}
-        return session
-
-    def _get_service(self):
+        if self._active_session:
+            use_active_session = True
+            decoded_token = jwt.decode(self._active_session.access_token_response.json()['access_token'],
+                                       options={'verify_signature': False})
+            expiration_time = datetime.fromtimestamp(decoded_token['exp'])
+            if expiration_time - datetime.utcnow() < timedelta(minutes=5):
+                LOG.debug('OAuth session expires at %s', expiration_time)
+                use_active_session = False
+            elif params is not None and 'scope' in params:
+                if sorted(decoded_token['scope']) != sorted(params['scope'].split(' ')):
+                    use_active_session = False
+                    LOG.debug('Scopes are not identical.')
+            if use_active_session:
+                return self._active_session
+
+        LOG.debug('Creating new OAuth session for "%s"', self.name)
+        if not params:
+            params = {'grant_type': 'client_credentials'}
         service = OAuth2Service(name=self.name, client_id=self.client_id, client_secret=self.client_secret,
                                 access_token_url=self.oauth_url)
-        return service
+        # the get_auth_session method of rauth does not check whether the response was 200 or not
+        # and therefore does not log a proper error message
+        self._active_session = service.get_auth_session('POST', data=params, decoder=json.loads)
```

Comment:
so could that result in `self._active_session` evaluating to true, but not having an `access_token_response` for example? And then causing an ugly (hard to trace) error eg in line 104 above?